### PR TITLE
Fix a bug in optimizer's mix_lr/max_lr when args.override_opt_param_scheduler==True

### PR DESCRIPTION
**Pull request by lyuwen**
1. In the current setting, OptimizerParamScheduler uses the `max_lr` and `min_lr` from optimizer instead of its own attributes to compute the learning rate at a certain step. 
2. When `--override-opt_param_scheduler` is used, user expects to override the learning rate schedule that is stored in the checkpoint. However, the optimizer still loads the `max_lr`, `min_lr` and the decoupled versions from the checkpoint. 
3. Therefore, when `--override-opt_param_scheduler` is used, the learning rate in training is computed using the old `max_lr` and the new `init_lr`, a mixture of old and new setting. This can be confusing for user to figure out the issue.

Here I propose an fix that during the `load_checkpoint`, when `--override-opt_param_scheduler` is set to `True`, the function `megatron.core.optimizer._update_min_and_max_lr_in_param_groups` is invoked to update the optimizer loaded from the checkpoint with the new learning rate boudaries. Thus the training would be carried out using the updated learning rate setting in situations like CPT, etc.

### CODE REVIEW

**Comment by lyuwen** [comment]
Hi @jon-barker , would you please have a look? It’s actually a bug that could affect tasks like continued pretraining.

**Review by dmarx — APPROVED** [review]
I've observed this behavior as well when trying to adjust the learning rate. Can confirm this fixes the bug. Modifying training hyper-parameters mid-training remains a headache though, e.g. offsetting to the current step via `--lr-warmup-iters` (don't even get me started on doing this with cosine decay)

**Comment by lyuwen** [comment]
@dmarx 
https://github.com/NVIDIA/Megatron-LM/blob/3235d439c9d76b62cf34108abb366013b4833cbf/megatron/core/optimizer_param_scheduler.py#L256-L264

Perhaps you want these to be wrapped around `self._check_and_set` so that they can be updated by the command line input?

**Comment by lyuwen** [comment]
If you use `--override-opt_param_scheduler` and set `--lr-warmup-iters` to the current step, wouldn't the training resume with the updated learning rate passed in from `--lr`?

**Comment by sbhavani** [comment]
Thanks @dmarx and @lyuwen! I've created an MR internally and will get back to you soon

**Comment by sbhavani** [comment]
@lyuwen Thanks for your contribution! As part of our new year cleanup, we're closing PRs opened before 2025.

This may still be relevant for checkpoint loading with overridden LR schedules. @NVIDIA/megatron-oncall 

If you'd like to pursue this fix, please feel free to submit a fresh PR. Thanks for your understanding!
